segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+        if (eventProcessorMap.containsKey(name)) {
+            return eventProcessorMap.get(name);
+        }
+
+        // Instantiate the EventProcessor and put it into the map.
+        DirectSegmentAccess segment = segmentSupplier.apply(name).join();
+        eventProcessorMap.put(name, new EventProcessorImpl(name, segment, handler, config));
+        return eventProcessorMap.get(name);
+    }
+
+    class EventProcessorImpl extends ContainerEventProcessor.EventProcessor {
+
+        // Maximum amount of bytes we allow to be non-processed before start rejecting add requests (4MBs).
+        private final static int MAX_OUTSTANDING_BYTES = 4 * 1024 * 1024;
+
+        private final DirectSegmentAccess segment;
+        private final AtomicLong outstandingBytes;
+        private final AtomicLong lastIterationLatency;
+        private final SegmentStoreMetrics.EventProcessor metrics;
+
+        public EventProcessorImpl(@NonNull String name, @NonNull DirectSegmentAccess segment,
+                                  @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                  @NonNull EventProcessorConfig config) {
+            super(name, handler, config);
+            this.segment = segment;
+            this.outstandingBytes = new AtomicLong(0);
+            this.lastIterationLatency = new AtomicLong(0);
+            this.metrics = new SegmentStoreMetrics.EventProcessor(name, containerId);
+        }
+
+        @Override
+        public CompletableFuture<Long> add(@NonNull BufferView event, Duration timeout) {
+            Preconditions.checkArgument(event.getLength() > 0);
+            Preconditions.checkState(outstandingBytes.get() < MAX_OUTSTANDING_BYTES,
+                    "Too many outstanding events for {}.", this.getName());
+
+            // Appends may be variable in size. We encode each event as [EVENT_LENGTH + EVENT_DATA]. This is needed to
+            // understand the boundaries of events in the absence of an actual notion of event.
+            return segment.append(BufferView.builder().add(intToByteArraySegment(event.getLength())).add(event).build(), null, timeout)
+                    .thenApply(offset -> outstandingBytes.addAndGet(event.getLength() + Integer.BYTES));
+        }
+
+        @Override
+        public void close() {
+            metrics.close();
+        }
+    }
+
+    @Override
+    public void close() {
+        log.info("Closing ContainerEventProcessor service.");
+        closed.set(true);
+    }
+
+    //endregion
+
+    //region AbstractService Implementation
+
+    @Override
+    protected void shutDown() throws Exception {
+        log.info("Shutting down ContainerEventProcessor service.");
+        eventProcessorMap.forEach((k, v) -> v.close());
+        eventProcessorMap.clear();
+        log.info("Shut down of ContainerEventProcessor service complete.");
+    }
+
+    @Override
+    protected void startUp() throws Exception {
```

Comment:
Consider following the pattern used in `StorageWriter`:
- Inherit from `AbstractThreadPoolService`
- Override the `run` method and return this `Futures.loop` from it.

Reply:
Done (it was `AbstractIdleService` before because issue #5909 suggested that).